api/src/main/java/org/eclipse/microprofile/faulttolerance/Asynchronous.java

```diff
@@ -29,16 +29,66 @@
 
 /**
  * Wrap the execution and invoke it asynchronously.
- * Any methods marked with this annotation must return {@link java.util.concurrent.Future}. 
+ * Any methods marked with this annotation must return one of:
+ * <ul>
+ *   <li>{@link java.util.concurrent.Future}</li>
+ *   <li>{@link java.util.concurrent.CompletionStage}</li>
+ * </ul>
  * Otherwise, {@link org.eclipse.microprofile.faulttolerance.exceptions.FaultToleranceDefinitionException} occurs.
- * Example usage:
  * 
+ * <p>
+ * When a method marked with this annotation is executed, the call returns immediately while the method
+ * is executed in another thread. The returned Future or CompletionStage is not completed until the result 
+ * of the method running asynchronously is completed (either normally or exceptionally) or the method 
+ * throws an exception.
+ * </p>
+ * 
+ * <p>After a method marked with this annotation returns normally, the Future or CompletionStage returned
+ * in the calling thread delegate all method calls to the object returned by the method. 
+ * CompletionStage returned in the calling thread is completed when the stage returned by the method is completed
+ * and with the same completion value or exception.</p>
+ * 
+ * <p>If a method marked with this annotation throws an exception when invoked, the exception isn't thrown 
+ * in the callers thread therefore it's not possible to catch it with a {@code try..catch} block. 
+ * The exception is propagated asynchronously:
+ * </p>
+ * <ul>
+ *   <li>If the method declares {@link java.util.concurrent.Future} as the return type, 
+ * calling {@link java.util.concurrent.Future#get()} will throw an 
+ * {@link java.util.concurrent.ExecutionException} wrapping the original exception</li>
+ *   <li>If the method declares {@link java.util.concurrent.CompletionStage} as the return type, 
+ * it is completed exceptionally with the exception.</li>
+ * </ul>
+ * It's recommended that methods throw only runtime exceptions to avoid unnecessary {@code try..catch} blocks.
+ * 
+ * <p>If a class is annotated with this annotation, all class methods that could be marked with this 
```

Comment:
I'm not sure this is clear enough. Reading this paragraph, I couldn't work out whether it would mean this class is a valid configuration or not:
```
@Asynchronous
public class MyClass {
   public String myMethodA() {}
   public Future<?> myMethodB() {}
}
```

I think this class should cause a definition exception to be thrown because `myMethodA` is not valid.

Can we just say:
> If a class is annotated with this annotation, all class methods are treated as if they were marked with this annotation.

If we really wanted to specify that methods that don't return `Future` or `CompletionStage` should be ignored, we could say
> If a class is annotated with this annotation, all class methods which return `Future` or `CompletionStage` are treated as if they were marked with this annotation.

I also couldn't see any TCKs that check validation of `@Asynchronous` methods or classes, so we should add some.

Reply:
I certainly had the second interpretation in mind, but now I can't see a usecase for mixing sync and async method where it would make sense to use `@Asynchronous` on the class level. 

Therefore I agree with:

> If a class is annotated with this annotation, all class methods are treated as if they were marked with this annotation.

And I would add:

> If one of the methods doesn't return Future/CompletionStage, {@link org.eclipse.microprofile.faulttolerance.exceptions.FaultToleranceDefinitionException} occurs (at deploy time if the bean is discovered during deployment). 


I would also add some tests to cover this.

If in a rare case most of methods are async and only a handful of methods are sync, `@Asynchronous` can still be added to all async methods. I think it's very rare and it's not worth to simplify at the expense of complexity.